File: java/src/test/java/sn/innolink/kyvshield/KyvShieldTest.java
package sn.innolink.kyvshield;

import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import sn.innolink.kyvshield.model.ChallengeMode;
import sn.innolink.kyvshield.model.ChallengesResponse;
import sn.innolink.kyvshield.model.KycResponse;
import sn.innolink.kyvshield.model.OverallStatus;
import sn.innolink.kyvshield.model.Step;
import sn.innolink.kyvshield.model.VerifyOptions;

import java.nio.charset.StandardCharsets;

/**
 * Integration tests for the KyvShield Java SDK.
 *
 * <p>These tests run against a locally started backend (port 8080) using the
 * standard demo API key and the sample CIN images bundled with the project.
 *
 * <p>Start the backend before running:
 * <pre>
 *   cd /Users/macbookpro/GolandProjects/cin_verification
 *   ./scripts/run-local.sh
 * </pre>
 *
 * <p>Image paths point to the test assets in the backend repository:
 * <ul>
 *   <li>{@code .../SN-CIN/RECTO.jpg}</li>
 *   <li>{@code .../SN-CIN/VERSO.jpg}</li>
 * </ul>
 */
public class KyvShieldTest {

    // ── Configuration ─────────────────────────────────────────────────────────

    private static final String LOCAL_BASE_URL = "http://localhost:8080";
    private static final String DEMO_API_KEY   = "kyvshield_demo_key_2024";

    private static final String RECTO_IMAGE =
            "/Users/macbookpro/GolandProjects/cin_verification/api/document_id/SN-CIN/RECTO.jpg";
    private static final String VERSO_IMAGE =
            "/Users/macbookpro/GolandProjects/cin_verification/api/document_id/SN-CIN/VERSO.jpg";

    // ── Setup ─────────────────────────────────────────────────────────────────

    private KyvShield kyv;

    @Before
    public void setUp() {
        kyv = new KyvShield(DEMO_API_KEY, LOCAL_BASE_URL);
    }

    // ── Constructor tests ─────────────────────────────────────────────────────

    @Test
    public void testConstructorDefaultUrl() {
        KyvShield client = new KyvShield(DEMO_API_KEY);
        Assert.assertNotNull(client);
    }

    @Test
    public void testConstructorCustomUrl() {
        KyvShield client = new KyvShield(DEMO_API_KEY, LOCAL_BASE_URL);
        Assert.assertNotNull(client);
    }

    @Test(expected = IllegalArgumentException.class)
    public void testConstructorNullApiKey() {
        new KyvShield(null);
    }

    @Test(expected = IllegalArgumentException.class)
    public void testConstructorBlankApiKey() {
        new KyvShield("   ");
    }

    // ── VerifyOptions builder tests ───────────────────────────────────────────

    @Test
    public void testVerifyOptionsBuilder() {
        VerifyOptions opts = new VerifyOptions.Builder()
                .steps(Step.RECTO, Step.VERSO)
                .target("SN-CIN")
                .language("fr")
                .challengeMode(ChallengeMode.STANDARD)
                .rectoChallengeMode(ChallengeMode.MINIMAL)
                .versoChallengeMode(ChallengeMode.MINIMAL)
                .requireFaceMatch(false)
                .kycIdentifier("test-user-123")
                .addImage("recto_center_document", RECTO_IMAGE)
                .addImage("verso_center_document", VERSO_IMAGE)
                .build();

        Assert.assertEquals(2, opts.getSteps().size());
        Assert.assertEquals(Step.RECTO, opts.getSteps().get(0));
        Assert.assertEquals(Step.VERSO, opts.getSteps().get(1));
        Assert.assertEquals("SN-CIN", opts.getTarget());
        Assert.assertEquals("fr", opts.getLanguage());
        Assert.assertEquals(ChallengeMode.STANDARD, opts.getChallengeMode());
        Assert.assertEquals(ChallengeMode.MINIMAL, opts.getRectoChallengeMode());
        Assert.assertEquals(ChallengeMode.MINIMAL, opts.getVersoChallengeMode());
        Assert.assertFalse(opts.isRequireFaceMatch());
        Assert.assertEquals("test-user-123", opts.getKycIdentifier());
        Assert.assertTrue(opts.getImages().containsKey("recto_center_document"));
        Assert.assertTrue(opts.getImages().containsKey("verso_center_document"));
    }

    @Test(expected = IllegalArgumentException.class)
    public void testVerifyOptionsBuilderMissingSteps() {
        new VerifyOptions.Builder()
                .target("SN-CIN")
                .addImage("recto_center_document", RECTO_IMAGE)
                .build();
    }

    @Test(expected = IllegalArgumentException.class)
    public void testVerifyOptionsBuilderMissingTarget() {
        new VerifyOptions.Builder()
                .steps(Step.RECTO)
                .addImage("recto_center_document", RECTO_IMAGE)
                .build();
    }

    @Test(expected = IllegalArgumentException.class)
    public void testVerifyOptionsBuilderMissingImages() {
        new VerifyOptions.Builder()
                .steps(Step.RECTO)
                .target("SN-CIN")
                .build();
    }

    // ── Enum tests ────────────────────────────────────────────────────────────

    @Test
    public void testChallengeModeValues() {
        Assert.assertEquals("minimal",  ChallengeMode.MINIMAL.getValue());
        Assert.assertEquals("standard", ChallengeMode.STANDARD.getValue());
        Assert.assertEquals("strict",   ChallengeMode.STRICT.getValue());
    }

    @Test
    public void testChallengeModeFromValue() {
        Assert.assertEquals(ChallengeMode.MINIMAL,  ChallengeMode.fromValue("minimal"));
        Assert.assertEquals(ChallengeMode.STANDARD, ChallengeMode.fromValue("STANDARD"));
        Assert.assertEquals(ChallengeMode.STRICT,   ChallengeMode.fromValue("strict"));
    }

    @Test(expected = IllegalArgumentException.class)
    public void testChallengeModeFromValueInvalid() {
        ChallengeMode.fromValue("unknown");
    }

    @Test
    public void testStepValues() {
        Assert.assertEquals("selfie", Step.SELFIE.getValue());
        Assert.assertEquals("recto",  Step.RECTO.getValue());
        Assert.assertEquals("verso",  Step.VERSO.getValue());
    }

    @Test
    public void testStepFromValue() {
        Assert.assertEquals(Step.SELFIE, Step.fromValue("selfie"));
        Assert.assertEquals(Step.RECTO,  Step.fromValue("RECTO"));
        Assert.assertEquals(Step.VERSO,  Step.fromValue("verso"));
    }

    @Test
    public void testOverallStatusFromValue() {
        Assert.assertEquals(OverallStatus.PASS,   OverallStatus.fromValue("pass"));
        Assert.assertEquals(OverallStatus.REJECT,  OverallStatus.fromValue("reject"));
        Assert.assertEquals(OverallStatus.REJECT,  OverallStatus.fromValue(null));
        Assert.assertEquals(OverallStatus.REJECT,  OverallStatus.fromValue("unknown"));
    }

    // ── Webhook signature tests ───────────────────────────────────────────────

    @Test
    public void testVerifyWebhookSignatureValid() {
        // The expected signature was computed offline with:
        //   echo -n "hello" | openssl dgst -sha256 -hmac "secret"
        // Output: 88aab3ede8d3adf94d26ab90d3bafd4a2083070c3bcce9c014ee04a443847c0b
        byte[] payload = "hello".getBytes(StandardCharsets.UTF_8);
        String expected = "88aab3ede8d3adf94d26ab90d3bafd4a2083070c3bcce9c014ee04a443847c0b";
        Assert.assertTrue(KyvShield.verifyWebhookSignature(payload, "secret", expected));
    }

    @Test
    public void testVerifyWebhookSignatureInvalid() {
        byte[] payload = "hello".getBytes(StandardCharsets.UTF_8);
        Assert.assertFalse(KyvShield.verifyWebhookSignature(payload, "secret", "badSignature"));
    }

    @Test
    public void testVerifyWebhookSignatureNullArgs() {
        Assert.assertFalse(KyvShield.verifyWebhookSignature(null, "key", "sig"));
        Assert.assertFalse(KyvShield.verifyWebhookSignature(new byte[0], null, "sig"));
        Assert.assertFalse(KyvShield.verifyWebhookSignature(new byte[0], "key", null));
    }

    // ── Integration tests (require running local backend) ─────────────────────

    /**
     * Calls {@code GET /api/v1/challenges} against the local backend.
     *
     * <p>The backend must be running at {@code http://localhost:8080}.
     * If it is not available the test is skipped rather than failed.
     */
    @Test
    public void testGetChallengesIntegration() {
        try {
            ChallengesResponse challenges = kyv.getChallenges();
            Assert.assertNotNull("challenges response must not be null", challenges);
            Assert.assertNotNull("document challenges must not be null", challenges.getDocument());
            Assert.assertNotNull("selfie challenges must not be null", challenges.getSelfie());
            System.out.println("[getChallenges] document.minimal  = "
                    + challenges.getDocument().getMinimal());
            System.out.println("[getChallenges] document.standard = "
                    + challenges.getDocument().getStandard());
            System.out.println("[getChallenges] selfie.standard   = "
                    + challenges.getSelfie().getStandard());
        } catch (KyvShieldException e) {
            System.out.println("[getChallenges] Backend unavailable, skipping: " + e.getMessage());
            // Not a hard failure when the backend is not running locally
        }
    }

    /**
     * Calls {@code POST /api/v1/kyc/verify} with the RECTO and VERSO sample images
     * against the local backend.
     *
     * <p>The backend must be running at {@code http://localhost:8080}.
     * If it is not available the test is skipped rather than failed.
     */
    @Test
    public void testVerifyRectoVersoIntegration() {
        VerifyOptions options = new VerifyOptions.Builder()
                .steps(Step.RECTO, Step.VERSO)
                .target("SN-CIN")
                .language("fr")
                .challengeMode(ChallengeMode.MINIMAL)
                .rectoChallengeMode(ChallengeMode.MINIMAL)
                .versoChallengeMode(ChallengeMode.MINIMAL)
                .requireFaceMatch(false)
                .addImage("recto_center_document", RECTO_IMAGE)
                .addImage("verso_center_document", VERSO_IMAGE)
                .build();

        try {
            KycResponse response = kyv.verify(options);

            Assert.assertNotNull("response must not be null", response);
            Assert.assertNotNull("sessionId must not be null", response.getSessionId());
            Assert.assertNotNull("overallStatus must not be null", response.getOverallStatus());
            Assert.assertNotNull("steps list must not be null", response.getSteps());
            Assert.assertEquals("steps count must match", 2, response.getSteps().size());

            System.out.println("[verify] sessionId       = " + response.getSessionId());
            System.out.println("[verify] overallStatus   = " + response.getOverallStatus());
            System.out.println("[verify] overallConfidence = " + response.getOverallConfidence());
            System.out.println("[verify] processingTimeMs  = " + response.getProcessingTimeMs());

            response.getSteps().forEach(step -> {
                System.out.println("[verify] step[" + step.getStepIndex() + "] "
                        + step.getStepType()
                        + " success=" + step.isSuccess()
                        + " processingTimeMs=" + step.getProcessingTimeMs());
                if (step.getLiveness() != null) {
                    System.out.println("  liveness: isLive=" + step.getLiveness().isLive()
                            + " score=" + step.getLiveness().getScore()
                            + " confidence=" + step.getLiveness().getConfidence());
                }
                if (step.getExtraction() != null && !step.getExtraction().isEmpty()) {
                    System.out.println("  extraction fields: " + step.getExtraction().size());
                    step.getExtraction().stream().limit(3).forEach(f ->
                            System.out.println("    " + f.getKey() + "=" + f.getValue()));
                }
            });

        } catch (KyvShieldException e) {
            if (e.getStatusCode() == 0) {
                // Network error — backend not running
                System.out.println("[verify] Backend unavailable, skipping: " + e.getMessage());
            } else {
                // HTTP error from the server — surface as test failure
                Assert.fail("Unexpected HTTP error " + e.getStatusCode()
                        + ": " + e.getMessage());
            }
        }
    }

    /**
     * Calls {@code POST /api/v1/kyc/verify} requesting a face match between the
     * selfie step and the document portrait.
     *
     * <p>Uses RECTO.jpg both as the document image and as a stand-in selfie to
     * avoid needing a separate selfie asset.
     */
    @Test
    public void testVerifyWithFaceMatchIntegration() {
        VerifyOptions options = new VerifyOptions.Builder()
                .steps(Step.SELFIE, Step.RECTO)
                .target("SN-CIN")
                .language("fr")
                .challengeMode(ChallengeMode.MINIMAL)
                .rectoChallengeMode(ChallengeMode.MINIMAL)
                .selfieChallengeMode(ChallengeMode.MINIMAL)
                .requireFaceMatch(true)
                .addImage("selfie_center_face", RECTO_IMAGE)       // stand-in selfie
                .addImage("selfie_close_eyes", RECTO_IMAGE)        // minimal selfie requires close_eyes too
                .addImage("recto_center_document", RECTO_IMAGE)
                .build();

        try {
            KycResponse response = kyv.verify(options);
            Assert.assertNotNull("response must not be null", response);

            if (response.getFaceVerification() != null) {
                System.out.println("[verify+face] isMatch=" + response.getFaceVerification().isMatch()
                        + " similarity=" + response.getFaceVerification().getSimilarityScore());
            } else {
                System.out.println("[verify+face] faceVerification not present in response");
            }

        } catch (KyvShieldException e) {
            if (e.getStatusCode() == 0) {
                System.out.println("[verify+face] Backend unavailable, skipping: " + e.getMessage());
            } else {
                Assert.fail("Unexpected HTTP error " + e.getStatusCode()
                        + ": " + e.getMessage());
            }
        }
    }

    /**
     * Verifies that the SDK throws {@link KyvShieldException} when an invalid
     * API key is provided.
     */
    @Test
    public void testInvalidApiKeyReturnsException() {
        KyvShield badClient = new KyvShield("invalid_key_xyz", LOCAL_BASE_URL);
        try {
            badClient.getChallenges();
            // Some backends return 200 for /challenges regardless of auth
        } catch (KyvShieldException e) {
            Assert.assertTrue("Should be an HTTP 401 or 403",
                    e.getStatusCode() == 401 || e.getStatusCode() == 403
                            || e.getStatusCode() == 0 /* network unavailable */);
        }
    }

    /**
     * Verifies that the SDK throws {@link KyvShieldException} when an image file
     * does not exist.
     */
    @Test(expected = KyvShieldException.class)
    public void testMissingImageFileThrows() {
        VerifyOptions options = new VerifyOptions.Builder()
                .steps(Step.RECTO)
                .target("SN-CIN")
                .addImage("recto_center_document", "/path/that/does/not/exist.jpg")
                .build();
        kyv.verify(options);
    }
}

File: kotlin/src/test/kotlin/sn/innolink/kyvshield/KyvShieldTest.kt
package sn.innolink.kyvshield

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import java.io.File
import java.nio.charset.StandardCharsets

/**
 * KyvShield Kotlin SDK — Integration Tests
 *
 * These tests run against a locally-running KyvShield backend at `http://localhost:8080`.
 *
 * Prerequisites:
 *   - Backend started via `./scripts/run-local.sh`
 *   - Set env var `KYVSHIELD_TEST_API_KEY` with a valid API key, or use the
 *     default demo key configured in Elasticsearch.
 *
 * Image fixtures:
 *   - Tests that exercise `verify()` need sample images.  Place them under:
 *       `src/test/resources/fixtures/`
 *   - Filenames expected:  `selfie.jpg`, `recto.jpg`, `verso.jpg`
 *
 * Run with:
 *   `./gradlew test`
 */
class KyvShieldTest {

    private val apiKey: String = System.getenv("KYVSHIELD_TEST_API_KEY") ?: "test_key_local"

    private val client = KyvShield(
        apiKey = apiKey,
        baseUrl = "http://localhost:8080",
        timeoutSeconds = 120,
    )

    // ─── getChallenges ────────────────────────────────────────────────────────

    @Test
    fun `getChallenges returns success and non-empty mode lists`() {
        val response = client.getChallenges()

        assertTrue(response.success, "Expected success=true from GET /api/v1/challenges")

        // Document challenges
        val doc = response.challenges.document
        assertTrue(doc.minimal.isNotEmpty(), "document.minimal should not be empty")
        assertTrue(doc.standard.isNotEmpty(), "document.standard should not be empty")
        assertTrue(doc.strict.isNotEmpty(), "document.strict should not be empty")

        // Selfie challenges
        val selfie = response.challenges.selfie
        assertTrue(selfie.minimal.isNotEmpty(), "selfie.minimal should not be empty")
        assertTrue(selfie.standard.isNotEmpty(), "selfie.standard should not be empty")
        assertTrue(selfie.strict.isNotEmpty(), "selfie.strict should not be empty")
    }

    @Test
    fun `getChallenges document minimal is a subset of standard`() {
        val response = client.getChallenges()
        val minimal = response.challenges.document.minimal.toSet()
        val standard = response.challenges.document.standard.toSet()
        // minimal challenges should all be present in standard
        assertTrue(standard.containsAll(minimal),
            "document.standard should include all document.minimal challenges")
    }

    // ─── verify — validation errors ──────────────────────────────────────────

    @Test
    fun `verify throws KyvShieldException on missing API key`() {
        val badClient = KyvShield(
            apiKey = "",
            baseUrl = "http://localhost:8080",
        )
        val options = minimalOptions()

        val ex = assertThrows<KyvShieldException> {
            badClient.verify(options)
        }
        assertNotNull(ex.statusCode, "Expected HTTP status code in exception")
        println("Got expected error: $ex")
    }

    @Test
    fun `verify throws KyvShieldException when image file does not exist`() {
        val options = VerifyOptions(
            steps = listOf(Step.RECTO),
            target = "SN-CIN",
            language = Language.ENGLISH,
            challengeMode = ChallengeMode.MINIMAL,
            images = mapOf("recto_center_document" to "/tmp/this_file_does_not_exist.jpg"),
        )

        val ex = assertThrows<KyvShieldException> {
            client.verify(options)
        }
        assertEquals("FILE_NOT_FOUND", ex.errorCode)
        println("Got expected FILE_NOT_FOUND: $ex")
    }

    // ─── verify — full recto-only (requires fixture) ─────────────────────────

    @Test
    fun `verify recto returns structured KycResponse`() {
        val rectoFile = fixtureFile("recto.jpg") ?: run {
            println("SKIP: src/test/resources/fixtures/recto.jpg not found")
            return
        }

        val options = VerifyOptions(
            steps = listOf(Step.RECTO),
            target = "SN-CIN",
            language = Language.FRENCH,
            challengeMode = ChallengeMode.MINIMAL,
            images = mapOf("recto_center_document" to rectoFile.absolutePath),
        )

        val response = client.verify(options)

        // Structure checks (not result checks — result depends on image quality)
        assertNotNull(response.sessionId, "sessionId must be non-null")
        assertTrue(response.sessionId.isNotBlank(), "sessionId must be non-blank")
        assertEquals(1, response.steps.size, "Expected exactly 1 step result")

        val step = response.steps[0]
        assertEquals(0, step.stepIndex)
        assertEquals(Step.RECTO, step.stepType)
        assertTrue(step.processingTimeMs >= 0)
        assertTrue(step.liveness.score in 0.0..1.0, "Liveness score must be 0–1")

        println("Recto verify: status=${response.overallStatus} confidence=${response.overallConfidence}")
    }

    @Test
    fun `verify selfie recto with face match returns face_verification block`() {
        val selfieFile = fixtureFile("selfie.jpg") ?: run {
            println("SKIP: src/test/resources/fixtures/selfie.jpg not found")
            return
        }
        val rectoFile = fixtureFile("recto.jpg") ?: run {
            println("SKIP: src/test/resources/fixtures/recto.jpg not found")
            return
        }

        val options = VerifyOptions(
            steps = listOf(Step.SELFIE, Step.RECTO),
            target = "SN-CIN",
            language = Language.ENGLISH,
            challengeMode = ChallengeMode.MINIMAL,
            requireFaceMatch = true,
            images = mapOf(
                "selfie_center_face"    to selfieFile.absolutePath,
                "recto_center_document" to rectoFile.absolutePath,
            ),
        )

        val response = client.verify(options)

        assertEquals(2, response.steps.size, "Expected 2 step results")
        assertNotNull(response.faceVerification, "face_verification must be present when requireFaceMatch=true")

        val fv = response.faceVerification!!
        assertTrue(fv.similarityScore >= 0.0, "similarity_score must be >= 0")
        println("Face match: isMatch=${fv.isMatch} score=${fv.similarityScore}")
    }

    @Test
    fun `verify full recto verso returns aligned_document and extraction`() {
        val rectoFile = fixtureFile("recto.jpg") ?: run {
            println("SKIP: fixtures not found")
            return
        }
        val versoFile = fixtureFile("verso.jpg") ?: run {
            println("SKIP: fixtures not found")
            return
        }

        val options = VerifyOptions(
            steps = listOf(Step.RECTO, Step.VERSO),
            target = "SN-CIN",
            language = Language.FRENCH,
            challengeMode = ChallengeMode.MINIMAL,
            images = mapOf(
                "recto_center_document" to rectoFile.absolutePath,
                "verso_center_document" to versoFile.absolutePath,
            ),
        )

        val response = client.verify(options)

        for (step in response.steps) {
            if (step.success) {
                assertNotNull(step.alignedDocument, "aligned_document expected on successful document step")
            }
        }
        println("Recto+Verso: status=${response.overallStatus}")
    }

    // ─── verify — per-step challenge mode override ────────────────────────────

    @Test
    fun `verify respects per-step challenge mode override`() {
        val rectoFile = fixtureFile("recto.jpg") ?: run {
            println("SKIP: fixtures not found")
            return
        }

        // Standard globally, but minimal for recto specifically
        val options = VerifyOptions(
            steps = listOf(Step.RECTO),
            target = "SN-CIN",
            language = Language.ENGLISH,
            challengeMode = ChallengeMode.STANDARD,
            stepChallengeModes = mapOf("recto" to ChallengeMode.MINIMAL),
            images = mapOf("recto_center_document" to rectoFile.absolutePath),
        )

        // If per-step override is respected, the server will only expect center_document
        // for recto (minimal mode). If standard were applied it would require more images
        // and return a 400 MISSING_FIELD — so a non-400 response validates the override.
        val response = client.verify(options)
        assertNotNull(response.sessionId)
        println("Per-step override test: status=${response.overallStatus}")
    }

    // ─── verify — kyc_identifier is echoed back ──────────────────────────────

    @Test
    fun `verify accepts kycIdentifier without error`() {
        val rectoFile = fixtureFile("recto.jpg") ?: run {
            println("SKIP: fixtures not found")
            return
        }

        val options = VerifyOptions(
            steps = listOf(Step.RECTO),
            target = "SN-CIN",
            language = Language.FRENCH,
            challengeMode = ChallengeMode.MINIMAL,
            kycIdentifier = "user-test-12345",
            images = mapOf("recto_center_document" to rectoFile.absolutePath),
        )

        val response = client.verify(options)
        // The session_id is a server-generated value; we just verify the call didn't blow up
        assertTrue(response.sessionId.isNotBlank())
        println("kycIdentifier test: sessionId=${response.sessionId}")
    }

    // ─── verifyWebhookSignature ───────────────────────────────────────────────

    @Test
    fun `verifyWebhookSignature accepts valid HMAC-SHA256 signature`() {
        val secret = "my_test_api_key"
        val payload = """{"event":"kyc.session.complete","session_id":"abc123"}"""
            .toByteArray(StandardCharsets.UTF_8)

        // Compute the expected signature using the same algorithm as the server
        val mac = javax.crypto.Mac.getInstance("HmacSHA256")
        mac.init(javax.crypto.spec.SecretKeySpec(secret.toByteArray(StandardCharsets.UTF_8), "HmacSHA256"))
        val expectedSig = mac.doFinal(payload).joinToString("") { "%02x".format(it) }

        val result = KyvShield.verifyWebhookSignature(
            payload = payload,
            apiKey = secret,
            signatureHeader = expectedSig,
        )
        assertTrue(result, "Valid HMAC-SHA256 signature should pass verification")
    }

    @Test
    fun `verifyWebhookSignature rejects tampered payload`() {
        val secret = "my_test_api_key"
        val payload = """{"event":"kyc.session.complete","session_id":"abc123"}"""
            .toByteArray(StandardCharsets.UTF_8)
        val tamperedPayload = """{"event":"kyc.session.complete","session_id":"EVIL999"}"""
            .toByteArray(StandardCharsets.UTF_8)

        val mac = javax.crypto.Mac.getInstance("HmacSHA256")
        mac.init(javax.crypto.spec.SecretKeySpec(secret.toByteArray(StandardCharsets.UTF_8), "HmacSHA256"))
        val sigForOriginal = mac.doFinal(payload).joinToString("") { "%02x".format(it) }

        val result = KyvShield.verifyWebhookSignature(
            payload = tamperedPayload,
            apiKey = secret,
            signatureHeader = sigForOriginal,
        )
        assertFalse(result, "Tampered payload should NOT pass signature verification")
    }

    @Test
    fun `verifyWebhookSignature rejects wrong API key`() {
        val secret = "correct_key"
        val wrongKey = "wrong_key"
        val payload = """{"session_id":"xyz"}""".toByteArray(StandardCharsets.UTF_8)

        val mac = javax.crypto.Mac.getInstance("HmacSHA256")
        mac.init(javax.crypto.spec.SecretKeySpec(secret.toByteArray(StandardCharsets.UTF_8), "HmacSHA256"))
        val sig = mac.doFinal(payload).joinToString("") { "%02x".format(it) }

        val result = KyvShield.verifyWebhookSignature(
            payload = payload,
            apiKey = wrongKey,
            signatureHeader = sig,
        )
        assertFalse(result, "Wrong API key should NOT pass signature verification")
    }

    // ─── Enum parsing ─────────────────────────────────────────────────────────

    @Test
    fun `OverallStatus fromString handles pass and reject`() {
        assertEquals(OverallStatus.PASS, OverallStatus.fromString("pass"))
        assertEquals(OverallStatus.REJECT, OverallStatus.fromString("reject"))
    }

    @Test
    fun `OverallStatus fromString throws on unknown value`() {
        assertThrows<IllegalArgumentException> {
            OverallStatus.fromString("unknown_status")
        }
    }

    @Test
    fun `Step fromString handles all valid values`() {
        assertEquals(Step.SELFIE, Step.fromString("selfie"))
        assertEquals(Step.RECTO, Step.fromString("recto"))
        assertEquals(Step.VERSO, Step.fromString("verso"))
    }

    @Test
    fun `ChallengeMode fromString handles all valid values`() {
        assertEquals(ChallengeMode.MINIMAL, ChallengeMode.fromString("minimal"))
        assertEquals(ChallengeMode.STANDARD, ChallengeMode.fromString("standard"))
        assertEquals(ChallengeMode.STRICT, ChallengeMode.fromString("strict"))
    }

    @Test
    fun `ConfidenceLevel fromString is case-insensitive`() {
        assertEquals(ConfidenceLevel.HIGH, ConfidenceLevel.fromString("HIGH"))
        assertEquals(ConfidenceLevel.HIGH, ConfidenceLevel.fromString("high"))
        assertEquals(ConfidenceLevel.MEDIUM, ConfidenceLevel.fromString("Medium"))
    }

    // ─── Helpers ──────────────────────────────────────────────────────────────

    /**
     * Builds a minimal [VerifyOptions] pointing to a non-existent image so that
     * the SDK throws FILE_NOT_FOUND before even sending the request.  Useful for
     * testing auth-failure paths by bypassing file-not-found errors.
     */
    private fun minimalOptions(): VerifyOptions {
        // Create a temporary file so the file check passes
        val tmpFile = File.createTempFile("kyvshield_test_", ".jpg").also {
            it.writeBytes(ByteArray(1024) { 0xFF.toByte() }) // minimal JPEG-ish bytes
            it.deleteOnExit()
        }
        return VerifyOptions(
            steps = listOf(Step.RECTO),
            target = "SN-CIN",
            language = Language.FRENCH,
            challengeMode = ChallengeMode.MINIMAL,
            images = mapOf("recto_center_document" to tmpFile.absolutePath),
        )
    }

    /**
     * Returns a [File] for a test fixture or `null` if it doesn't exist.
     * Fixtures live in `src/test/resources/fixtures/`.
     */
    private fun fixtureFile(name: String): File? {
        val paths = listOf(
            "src/test/resources/fixtures/$name",
            "kotlin/src/test/resources/fixtures/$name",
        )
        return paths.map { File(it) }.firstOrNull { it.exists() && it.isFile }
    }
}
